modules/kafka/src/main/java/org/testcontainers/containers/KafkaContainer.java

```diff
@@ -44,80 +53,134 @@ public KafkaContainer(final DockerImageName dockerImageName) {
         return self();
     }
 
+    public KafkaContainer withKraft() {
+        if (this.externalZookeeperConnect != null) {
+            throw new IllegalStateException("Cannot configure Kraft mode when Zookeeper configured");
+        }
+        String actualVersion = DockerImageName.parse(getDockerImageName()).getVersionPart();
+        if (new ComparableVersion(actualVersion).isLessThan(MIN_KRAFT_TAG)) {
+            throw new IllegalArgumentException(
+                String.format(
+                    "Provided Confluent Platform's version %s is not supported in Kraft mode (must be %s or above)",
+                    actualVersion,
+                    MIN_KRAFT_TAG
+                )
+            );
+        }
+        kraftEnabled = true;
+        return self();
+    }
+
     public String getBootstrapServers() {
         return String.format("PLAINTEXT://%s:%s", getHost(), getMappedPort(KAFKA_PORT));
     }
 
     @Override
     protected void configure() {
-        withEnv(
-            "KAFKA_ADVERTISED_LISTENERS",
-            String.format("BROKER://%s:9092", getNetwork() != null ? getNetworkAliases().get(0) : "localhost")
-        );
+        if (kraftEnabled) {
+            waitingFor(Wait.forLogMessage(".*Transitioning from RECOVERY to RUNNING.*", 1));
+        } else {
+            waitingFor(Wait.forLogMessage(".*\\[KafkaServer id=\\d+\\] started.*", 1));
+        }
+        // Use two listeners with different names, it will force Kafka to communicate with itself via internal
+        // listener when KAFKA_INTER_BROKER_LISTENER_NAME is set, otherwise Kafka will try to use the advertised listener
+        withEnv("KAFKA_LISTENERS", "PLAINTEXT://0.0.0.0:" + KAFKA_PORT + ",BROKER://0.0.0.0:9092");
+        withEnv("KAFKA_LISTENER_SECURITY_PROTOCOL_MAP", "BROKER:PLAINTEXT,PLAINTEXT:PLAINTEXT");
+        withEnv("KAFKA_INTER_BROKER_LISTENER_NAME", "BROKER");
 
-        String command = "#!/bin/bash\n";
-        if (externalZookeeperConnect != null) {
+        withEnv("KAFKA_BROKER_ID", "1");
+        withEnv("KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR", DEFAULT_INTERNAL_TOPIC_RF);
+        withEnv("KAFKA_OFFSETS_TOPIC_NUM_PARTITIONS", DEFAULT_INTERNAL_TOPIC_RF);
+        withEnv("KAFKA_TRANSACTION_STATE_LOG_REPLICATION_FACTOR", DEFAULT_INTERNAL_TOPIC_RF);
+        withEnv("KAFKA_TRANSACTION_STATE_LOG_MIN_ISR", DEFAULT_INTERNAL_TOPIC_RF);
+        withEnv("KAFKA_LOG_FLUSH_INTERVAL_MESSAGES", Long.MAX_VALUE + "");
+        withEnv("KAFKA_GROUP_INITIAL_REBALANCE_DELAY_MS", "0");
+
+        if (kraftEnabled) {
+            withEnv(
+                "KAFKA_LISTENER_SECURITY_PROTOCOL_MAP",
+                String.format("%s,CONTROLLER:PLAINTEXT", getEnvMap().get("KAFKA_LISTENER_SECURITY_PROTOCOL_MAP"))
+            );
+            withEnv(
+                "KAFKA_LISTENERS",
+                String.format("%s,CONTROLLER://0.0.0.0:9094", getEnvMap().get("KAFKA_LISTENERS"))
+            );
+
+            withEnv("KAFKA_NODE_ID", "1");
```

Comment:
I think we will have the same issue as `broker.id` if building a KafkaCluster with kraft mode. See [node.id](https://docs.confluent.io/platform/current/installation/configuration/broker-configs.html#brokerconfigs_node.id)

I think would be great to have a similar example for in Kraft mode. (if possible)

Reply:
Yes, I need to reconsider the brokers configuration.

> I think would be great to have a similar example for in Kraft mode. (if possible)
That's the plan!